### ISSUE TITLE
Warn about unreachable code in flow control statements

### ISSUE THREAD

**Issue by ritiek** [rust-lang/rust#47386]
For example, this code compiles without warning:
```rust
fn foo() -> u32 {
    let possible = false;
    if possible {
        return 0;
    }
    1
}
```
Here, `possible` has no way of becoming *true* and therefore anything under `if possible` will never be executed.

I think the compiler should warn about such unreachable code. Something similar to what it does with:

```rust
fn foo() -> u32 {
    return 0;
    return 1;
```
```
warning: unreachable statement
 --> test.rs:3:5
  |
3 |     return 1;
  |     ^^^^^^^^^
  |
```

What do you guys think?

**Comment by nagisa** [comment]
We shouldn’t do that because

```
if cfg!(...) {
}
```

is a fairly common pattern and the thing you’re proposing would mean many false-positives. Adding new lints is something that requires an RFC, so I'm inclined to judge that doing such a potentially controversial change should also require an RFC.

**Comment by ritiek** [comment]
I see. It will indeed introduce many false positives, probably not going to be worth the effort either when we already have `cfg!`.
